{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "29af349f-923d-4e0d-a9da-b5f798653da6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import category_encoders as ce\n",
    "import logging\n",
    "import os\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.feature_selection import VarianceThreshold\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Create a logger\n",
    "logging.basicConfig(format=\"%(asctime)s %(levelname)s: %(message)s\")\n",
    "logger = logging.getLogger()\n",
    "logger.setLevel(logging.INFO)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a6d19d7-0092-4db5-8f0a-478960a424c6",
   "metadata": {},
   "source": [
    "### Constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0b8b32e4-5d38-4feb-bca4-51cefd2dbaf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directories\n",
    "processed_dir = \"data\"\n",
    "\n",
    "sample_sheets_dir = \"sample_sheets\"\n",
    "\n",
    "cnv_data_dir = \"/mnt/d/Documents/Data/TCGA/CNV\"\n",
    "dna_methylation_data_dir = \"/mnt/d/Documents/Data/TCGA/DNA_Methylation\"\n",
    "gene_expression_data_dir = \"/mnt/d/Documents/Data/TCGA/gene_expression\"\n",
    "mirna_data_dir = \"/mnt/d/Documents/Data/TCGA/miRNA\"\n",
    "\n",
    "# Statuses\n",
    "cnv_process_status = \"processing\"\n",
    "dna_methylation_process_status = \"processing\"\n",
    "gene_expression_process_status = \"processing\"\n",
    "miRNA_process_status = \"processing\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "779aff46-3b25-41fe-bdca-85e8cb3df04b",
   "metadata": {},
   "source": [
    "### Preprocess clinical data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bd7bb91f-2dbc-4e62-ad1f-5db20fb574aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clinical = pd.read_csv(f\"{sample_sheets_dir}/gdc_clinical.tsv\", sep=\"\\t\")\n",
    "# df.columns.to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "341af2ae-dab2-43a7-a540-c09ac6857fda",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-01-14 13:53:42,929 INFO: (1107, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_submitter_id</th>\n",
       "      <th>age_at_index</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>days_to_last_follow_up</th>\n",
       "      <th>morphology</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>ajcc_pathologic_m</th>\n",
       "      <th>ajcc_pathologic_n</th>\n",
       "      <th>ajcc_pathologic_stage</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>treatment_or_therapy</th>\n",
       "      <th>treatment_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-62-A471</td>\n",
       "      <td>64</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1246.0</td>\n",
       "      <td>8140/3</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>Alive</td>\n",
       "      <td>M0</td>\n",
       "      <td>N1</td>\n",
       "      <td>Stage IIB</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>yes</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-62-A471</td>\n",
       "      <td>64</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1246.0</td>\n",
       "      <td>8140/3</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>Alive</td>\n",
       "      <td>M0</td>\n",
       "      <td>N1</td>\n",
       "      <td>Stage IIB</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>no</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-67-3773</td>\n",
       "      <td>84</td>\n",
       "      <td>NaN</td>\n",
       "      <td>427.0</td>\n",
       "      <td>8140/3</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>Alive</td>\n",
       "      <td>M0</td>\n",
       "      <td>N0</td>\n",
       "      <td>Stage IB</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>not reported</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-67-3773</td>\n",
       "      <td>84</td>\n",
       "      <td>NaN</td>\n",
       "      <td>427.0</td>\n",
       "      <td>8140/3</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>Alive</td>\n",
       "      <td>M0</td>\n",
       "      <td>N0</td>\n",
       "      <td>Stage IB</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>not reported</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-17-Z038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1102</th>\n",
       "      <td>TCGA-55-7570</td>\n",
       "      <td>60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>824.0</td>\n",
       "      <td>8140/3</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>male</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>Alive</td>\n",
       "      <td>MX</td>\n",
       "      <td>N0</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>no</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103</th>\n",
       "      <td>TCGA-78-7146</td>\n",
       "      <td>71</td>\n",
       "      <td>173</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8255/3</td>\n",
       "      <td>not reported</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>Dead</td>\n",
       "      <td>M0</td>\n",
       "      <td>N2</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>Adenocarcinoma with mixed subtypes</td>\n",
       "      <td>no</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1104</th>\n",
       "      <td>TCGA-78-7146</td>\n",
       "      <td>71</td>\n",
       "      <td>173</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8255/3</td>\n",
       "      <td>not reported</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>Dead</td>\n",
       "      <td>M0</td>\n",
       "      <td>N2</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>Adenocarcinoma with mixed subtypes</td>\n",
       "      <td>no</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105</th>\n",
       "      <td>TCGA-44-3398</td>\n",
       "      <td>77</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1163.0</td>\n",
       "      <td>8140/3</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>Alive</td>\n",
       "      <td>M0</td>\n",
       "      <td>N0</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>no</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1106</th>\n",
       "      <td>TCGA-44-3398</td>\n",
       "      <td>77</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1163.0</td>\n",
       "      <td>8140/3</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>Alive</td>\n",
       "      <td>M0</td>\n",
       "      <td>N0</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>no</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1107 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     case_submitter_id age_at_index days_to_death days_to_last_follow_up  \\\n",
       "0         TCGA-62-A471           64           NaN                 1246.0   \n",
       "1         TCGA-62-A471           64           NaN                 1246.0   \n",
       "2         TCGA-67-3773           84           NaN                  427.0   \n",
       "3         TCGA-67-3773           84           NaN                  427.0   \n",
       "4         TCGA-17-Z038          NaN           NaN                    NaN   \n",
       "...                ...          ...           ...                    ...   \n",
       "1102      TCGA-55-7570           60           NaN                  824.0   \n",
       "1103      TCGA-78-7146           71           173                    NaN   \n",
       "1104      TCGA-78-7146           71           173                    NaN   \n",
       "1105      TCGA-44-3398           77           NaN                 1163.0   \n",
       "1106      TCGA-44-3398           77           NaN                 1163.0   \n",
       "\n",
       "     morphology               ethnicity  gender                       race  \\\n",
       "0        8140/3  not hispanic or latino    male                      white   \n",
       "1        8140/3  not hispanic or latino    male                      white   \n",
       "2        8140/3  not hispanic or latino  female                      white   \n",
       "3        8140/3  not hispanic or latino  female                      white   \n",
       "4           NaN                     NaN     NaN                        NaN   \n",
       "...         ...                     ...     ...                        ...   \n",
       "1102     8140/3  not hispanic or latino    male  black or african american   \n",
       "1103     8255/3            not reported  female                      white   \n",
       "1104     8255/3            not reported  female                      white   \n",
       "1105     8140/3  not hispanic or latino  female                      white   \n",
       "1106     8140/3  not hispanic or latino  female                      white   \n",
       "\n",
       "     vital_status ajcc_pathologic_m ajcc_pathologic_n ajcc_pathologic_stage  \\\n",
       "0           Alive                M0                N1             Stage IIB   \n",
       "1           Alive                M0                N1             Stage IIB   \n",
       "2           Alive                M0                N0              Stage IB   \n",
       "3           Alive                M0                N0              Stage IB   \n",
       "4             NaN               NaN               NaN                   NaN   \n",
       "...           ...               ...               ...                   ...   \n",
       "1102        Alive                MX                N0              Stage IA   \n",
       "1103         Dead                M0                N2            Stage IIIA   \n",
       "1104         Dead                M0                N2            Stage IIIA   \n",
       "1105        Alive                M0                N0              Stage IA   \n",
       "1106        Alive                M0                N0              Stage IA   \n",
       "\n",
       "                       primary_diagnosis treatment_or_therapy  \\\n",
       "0                    Adenocarcinoma, NOS                  yes   \n",
       "1                    Adenocarcinoma, NOS                   no   \n",
       "2                    Adenocarcinoma, NOS         not reported   \n",
       "3                    Adenocarcinoma, NOS         not reported   \n",
       "4                                    NaN                  NaN   \n",
       "...                                  ...                  ...   \n",
       "1102                 Adenocarcinoma, NOS                   no   \n",
       "1103  Adenocarcinoma with mixed subtypes                   no   \n",
       "1104  Adenocarcinoma with mixed subtypes                   no   \n",
       "1105                 Adenocarcinoma, NOS                   no   \n",
       "1106                 Adenocarcinoma, NOS                   no   \n",
       "\n",
       "                   treatment_type  \n",
       "0     Pharmaceutical Therapy, NOS  \n",
       "1          Radiation Therapy, NOS  \n",
       "2          Radiation Therapy, NOS  \n",
       "3     Pharmaceutical Therapy, NOS  \n",
       "4                             NaN  \n",
       "...                           ...  \n",
       "1102  Pharmaceutical Therapy, NOS  \n",
       "1103       Radiation Therapy, NOS  \n",
       "1104  Pharmaceutical Therapy, NOS  \n",
       "1105  Pharmaceutical Therapy, NOS  \n",
       "1106       Radiation Therapy, NOS  \n",
       "\n",
       "[1107 rows x 15 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select only some fields\n",
    "selected_cols = [\n",
    "    \"case_submitter_id\",\n",
    "    \"age_at_index\",\n",
    "    \"days_to_death\",\n",
    "    \"days_to_last_follow_up\",\n",
    "    \"morphology\",\n",
    "    \"ethnicity\",\n",
    "    \"gender\",\n",
    "    \"race\",\n",
    "    \"vital_status\",\n",
    "    # \"year_of_birth\",\n",
    "    # \"year_of_death\",\n",
    "    \"ajcc_pathologic_m\",\n",
    "    \"ajcc_pathologic_n\",\n",
    "    \"ajcc_pathologic_stage\",\n",
    "    \"primary_diagnosis\",\n",
    "    \"treatment_or_therapy\",\n",
    "    \"treatment_type\",\n",
    "]\n",
    "\n",
    "df_clinical = df_clinical[selected_cols]\n",
    "\n",
    "# Replace '--' with NaN\n",
    "df_clinical = df_clinical.replace(\"'--\", np.nan)\n",
    "\n",
    "logger.info(df_clinical.shape)\n",
    "df_clinical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "69635ca0-51a9-4e45-a885-d583497d0361",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add new column for Pharmaceutical Therapy, NOS\n",
    "# Add new column for Radiation Therapy, NOS\n",
    "df_clinical['pharmaceutical_treatment'] = ((df_clinical['treatment_type'] == 'Pharmaceutical Therapy, NOS') & (df_clinical[\"treatment_or_therapy\"] == \"yes\")).astype(int)\n",
    "df_clinical['radiation_treatment'] = ((df_clinical['treatment_type'] == 'Radiation Therapy, NOS') & (df_clinical[\"treatment_or_therapy\"] == \"yes\")).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "af58aa9b-e7fc-4798-b382-3eff41eb9c8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-01-14 13:53:42,964 INFO: Total samples: 509\n",
      "2025-01-14 13:53:42,964 INFO: Total clincal features: 16\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>age_at_index</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>days_to_last_follow_up</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>morphology</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>ajcc_pathologic_m</th>\n",
       "      <th>ajcc_pathologic_n</th>\n",
       "      <th>ajcc_pathologic_stage</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>treatment_type</th>\n",
       "      <th>event</th>\n",
       "      <th>days_to_event</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-05-4245</td>\n",
       "      <td>81</td>\n",
       "      <td>None</td>\n",
       "      <td>730.0</td>\n",
       "      <td>not reported</td>\n",
       "      <td>male</td>\n",
       "      <td>not reported</td>\n",
       "      <td>8140/3</td>\n",
       "      <td>Alive</td>\n",
       "      <td>M0</td>\n",
       "      <td>N2</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>730.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-05-4249</td>\n",
       "      <td>67</td>\n",
       "      <td>None</td>\n",
       "      <td>1523.0</td>\n",
       "      <td>not reported</td>\n",
       "      <td>male</td>\n",
       "      <td>not reported</td>\n",
       "      <td>8140/3</td>\n",
       "      <td>Alive</td>\n",
       "      <td>M0</td>\n",
       "      <td>N0</td>\n",
       "      <td>Stage IB</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>1523.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-05-4250</td>\n",
       "      <td>79</td>\n",
       "      <td>121</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>8140/3</td>\n",
       "      <td>Dead</td>\n",
       "      <td>M0</td>\n",
       "      <td>N1</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>121.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-05-4382</td>\n",
       "      <td>68</td>\n",
       "      <td>None</td>\n",
       "      <td>607.0</td>\n",
       "      <td>not reported</td>\n",
       "      <td>male</td>\n",
       "      <td>not reported</td>\n",
       "      <td>8255/3</td>\n",
       "      <td>Alive</td>\n",
       "      <td>M0</td>\n",
       "      <td>N0</td>\n",
       "      <td>Stage IB</td>\n",
       "      <td>Adenocarcinoma with mixed subtypes</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "      <td>0</td>\n",
       "      <td>607.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TCGA-05-4384</td>\n",
       "      <td>66</td>\n",
       "      <td>None</td>\n",
       "      <td>426.0</td>\n",
       "      <td>not reported</td>\n",
       "      <td>male</td>\n",
       "      <td>not reported</td>\n",
       "      <td>8255/3</td>\n",
       "      <td>Alive</td>\n",
       "      <td>M0</td>\n",
       "      <td>N2</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>Adenocarcinoma with mixed subtypes</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "      <td>0</td>\n",
       "      <td>426.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>TCGA-NJ-A55O</td>\n",
       "      <td>56</td>\n",
       "      <td>None</td>\n",
       "      <td>13.0</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>8480/3</td>\n",
       "      <td>Alive</td>\n",
       "      <td>M0</td>\n",
       "      <td>N1</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>Mucinous adenocarcinoma</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581</th>\n",
       "      <td>TCGA-NJ-A55R</td>\n",
       "      <td>67</td>\n",
       "      <td>None</td>\n",
       "      <td>603.0</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>8230/3</td>\n",
       "      <td>Alive</td>\n",
       "      <td>MX</td>\n",
       "      <td>N0</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>Solid carcinoma, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>603.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>582</th>\n",
       "      <td>TCGA-NJ-A7XG</td>\n",
       "      <td>49</td>\n",
       "      <td>None</td>\n",
       "      <td>617.0</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>male</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>8140/3</td>\n",
       "      <td>Alive</td>\n",
       "      <td>M0</td>\n",
       "      <td>N1</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>0</td>\n",
       "      <td>617.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>TCGA-O1-A52J</td>\n",
       "      <td>74</td>\n",
       "      <td>1798</td>\n",
       "      <td>None</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>8140/3</td>\n",
       "      <td>Dead</td>\n",
       "      <td>MX</td>\n",
       "      <td>N0</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>1798.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>TCGA-S2-AA1A</td>\n",
       "      <td>68</td>\n",
       "      <td>None</td>\n",
       "      <td>513.0</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>8253/3</td>\n",
       "      <td>Alive</td>\n",
       "      <td>M0</td>\n",
       "      <td>N0</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>Bronchio-alveolar carcinoma, mucinous</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>513.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>509 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          case_id age_at_index days_to_death days_to_last_follow_up  \\\n",
       "1    TCGA-05-4245           81          None                  730.0   \n",
       "2    TCGA-05-4249           67          None                 1523.0   \n",
       "3    TCGA-05-4250           79           121                   None   \n",
       "4    TCGA-05-4382           68          None                  607.0   \n",
       "5    TCGA-05-4384           66          None                  426.0   \n",
       "..            ...          ...           ...                    ...   \n",
       "580  TCGA-NJ-A55O           56          None                   13.0   \n",
       "581  TCGA-NJ-A55R           67          None                  603.0   \n",
       "582  TCGA-NJ-A7XG           49          None                  617.0   \n",
       "583  TCGA-O1-A52J           74          1798                   None   \n",
       "584  TCGA-S2-AA1A           68          None                  513.0   \n",
       "\n",
       "                  ethnicity  gender                       race morphology  \\\n",
       "1              not reported    male               not reported     8140/3   \n",
       "2              not reported    male               not reported     8140/3   \n",
       "3              not reported  female               not reported     8140/3   \n",
       "4              not reported    male               not reported     8255/3   \n",
       "5              not reported    male               not reported     8255/3   \n",
       "..                      ...     ...                        ...        ...   \n",
       "580  not hispanic or latino  female                      white     8480/3   \n",
       "581  not hispanic or latino    male                      white     8230/3   \n",
       "582  not hispanic or latino    male  black or african american     8140/3   \n",
       "583  not hispanic or latino  female                      white     8140/3   \n",
       "584  not hispanic or latino  female  black or african american     8253/3   \n",
       "\n",
       "    vital_status ajcc_pathologic_m ajcc_pathologic_n ajcc_pathologic_stage  \\\n",
       "1          Alive                M0                N2            Stage IIIA   \n",
       "2          Alive                M0                N0              Stage IB   \n",
       "3           Dead                M0                N1            Stage IIIA   \n",
       "4          Alive                M0                N0              Stage IB   \n",
       "5          Alive                M0                N2            Stage IIIA   \n",
       "..           ...               ...               ...                   ...   \n",
       "580        Alive                M0                N1             Stage IIA   \n",
       "581        Alive                MX                N0              Stage IA   \n",
       "582        Alive                M0                N1            Stage IIIA   \n",
       "583         Dead                MX                N0              Stage IA   \n",
       "584        Alive                M0                N0              Stage IA   \n",
       "\n",
       "                         primary_diagnosis               treatment_type  \\\n",
       "1                      Adenocarcinoma, NOS                         None   \n",
       "2                      Adenocarcinoma, NOS                         None   \n",
       "3                      Adenocarcinoma, NOS                         None   \n",
       "4       Adenocarcinoma with mixed subtypes       Radiation Therapy, NOS   \n",
       "5       Adenocarcinoma with mixed subtypes       Radiation Therapy, NOS   \n",
       "..                                     ...                          ...   \n",
       "580                Mucinous adenocarcinoma                         None   \n",
       "581                   Solid carcinoma, NOS                         None   \n",
       "582                    Adenocarcinoma, NOS  Pharmaceutical Therapy, NOS   \n",
       "583                    Adenocarcinoma, NOS                         None   \n",
       "584  Bronchio-alveolar carcinoma, mucinous                         None   \n",
       "\n",
       "     event  days_to_event  \n",
       "1        0          730.0  \n",
       "2        0         1523.0  \n",
       "3        1          121.0  \n",
       "4        0          607.0  \n",
       "5        0          426.0  \n",
       "..     ...            ...  \n",
       "580      0           13.0  \n",
       "581      0          603.0  \n",
       "582      0          617.0  \n",
       "583      1         1798.0  \n",
       "584      0          513.0  \n",
       "\n",
       "[509 rows x 16 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group by 'case_submitter_id' and aggregate\n",
    "df_clinical = df_clinical.groupby('case_submitter_id').agg({\n",
    "    \"age_at_index\": \"first\",\n",
    "    \"days_to_death\": \"first\",\n",
    "    \"days_to_last_follow_up\": \"first\",\n",
    "    \"ethnicity\": \"first\",\n",
    "    \"gender\": \"first\",\n",
    "    \"race\": \"first\",\n",
    "    \"age_at_index\": \"first\",\n",
    "    \"morphology\": \"first\",\n",
    "    \"vital_status\": \"first\",\n",
    "    \"ajcc_pathologic_m\": \"first\",\n",
    "    \"ajcc_pathologic_n\": \"first\",\n",
    "    \"ajcc_pathologic_stage\": \"first\",\n",
    "    \"primary_diagnosis\": \"first\",\n",
    "    \"pharmaceutical_treatment\": \"max\",             # Update to 1 if any row has 1\n",
    "    \"radiation_treatment\": \"max\"                   # Update to 1 if any row has 1\n",
    "}).reset_index()\n",
    "\n",
    "# Update 'treatment_type' based on 'treatment_or_therapy'\n",
    "df_clinical['treatment_type'] = np.where(df_clinical['pharmaceutical_treatment'] == 1, 'Pharmaceutical Therapy, NOS', 'None')\n",
    "df_clinical['treatment_type'] = np.where(df_clinical['radiation_treatment'] == 1, 'Radiation Therapy, NOS', df_clinical['treatment_type'])\n",
    "\n",
    "# Remove rows where 'vital_status' is NaN\n",
    "df_clinical = df_clinical.dropna(subset=['vital_status'])\n",
    "\n",
    "# Remove rows where 'days_to_death' or 'days_to_last_follow_up' is NaN\n",
    "df_clinical = df_clinical[(df_clinical[\"days_to_death\"].notna()) | (df_clinical[\"days_to_last_follow_up\"].notna())]\n",
    "\n",
    "# Convert vital_status to a binary event indicator (1 = Dead, 0 = Alive)\n",
    "df_clinical['event'] = (df_clinical['vital_status'] == 'Dead').astype(int)\n",
    "\n",
    "# Get 'days_to_event'\n",
    "df_clinical['days_to_event'] = np.where((df_clinical['days_to_death'].notna()) & (df_clinical['event'] == 1), df_clinical['days_to_death'], df_clinical['days_to_last_follow_up']).astype(float)\n",
    "\n",
    "df_clinical = df_clinical[df_clinical['days_to_event'] > 0]\n",
    "\n",
    "# Rename some columns\n",
    "df_clinical = df_clinical.rename(columns={'case_submitter_id': 'case_id'})\n",
    "\n",
    "# Drop some columns\n",
    "df_clinical = df_clinical.drop(columns=['pharmaceutical_treatment', 'radiation_treatment'])\n",
    "\n",
    "logger.info(f\"Total samples: {df_clinical.shape[0]}\")\n",
    "logger.info(f\"Total clincal features: { df_clinical.shape[1]}\")\n",
    "\n",
    "df_clinical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bc033367-495e-4a44-a102-f13e13a0e716",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export data\n",
    "df_clinical.to_csv(f\"{processed_dir}/clincal.tsv\", sep=\"\\t\")\n",
    "\n",
    "# Get list of cases\n",
    "list_cases = df_clinical[\"case_id\"].to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0059b44-54de-4a1f-aee8-8bb93ffef783",
   "metadata": {},
   "source": [
    "### Preprocess CNV data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d89ce397-9c61-4816-8b93-83626f0967d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cnv_sample_sheet = pd.read_csv(f\"{sample_sheets_dir}/gdc_cnv_sample_sheet.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7a163280-9161-4091-8dcf-1036be2ae481",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File ID</th>\n",
       "      <th>File Name</th>\n",
       "      <th>Data Category</th>\n",
       "      <th>Data Type</th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Case ID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Sample Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>95e5fe95-789d-4223-b9cc-a16c7fd683c1</td>\n",
       "      <td>TCGA-LUAD.6ab7dce7-f54d-4f9e-b232-7f094e8bf05d...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Gene Level Copy Number</td>\n",
       "      <td>TCGA-LUAD</td>\n",
       "      <td>TCGA-38-4630</td>\n",
       "      <td>TCGA-38-4630-01A</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8c9ac2ea-6123-445c-8bd4-b0ea7c681edf</td>\n",
       "      <td>TCGA-LUAD.6e3bfcae-bea1-46b3-afee-df2e3a344cd1...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Gene Level Copy Number</td>\n",
       "      <td>TCGA-LUAD</td>\n",
       "      <td>TCGA-50-5932</td>\n",
       "      <td>TCGA-50-5932-01A</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b09d319f-7fba-4677-9aa2-fb2666643884</td>\n",
       "      <td>TCGA-LUAD.1afb48ce-59a8-4130-8ff6-45d29405087e...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Gene Level Copy Number</td>\n",
       "      <td>TCGA-LUAD</td>\n",
       "      <td>TCGA-J2-A4AD</td>\n",
       "      <td>TCGA-J2-A4AD-01A</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>74cb107f-adc9-4aee-8381-c95831dbd4ea</td>\n",
       "      <td>TCGA-LUAD.d0597d14-df0c-413b-888f-53597d1fb61a...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Gene Level Copy Number</td>\n",
       "      <td>TCGA-LUAD</td>\n",
       "      <td>TCGA-38-4625</td>\n",
       "      <td>TCGA-38-4625-01A</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b6b74adb-31e0-4ecf-9a65-aa8a1babc46b</td>\n",
       "      <td>TCGA-LUAD.780c0b4e-e589-4cc3-b0da-ea99e908afdd...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Gene Level Copy Number</td>\n",
       "      <td>TCGA-LUAD</td>\n",
       "      <td>TCGA-55-7284</td>\n",
       "      <td>TCGA-55-7284-01B</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                File ID  \\\n",
       "0  95e5fe95-789d-4223-b9cc-a16c7fd683c1   \n",
       "1  8c9ac2ea-6123-445c-8bd4-b0ea7c681edf   \n",
       "2  b09d319f-7fba-4677-9aa2-fb2666643884   \n",
       "3  74cb107f-adc9-4aee-8381-c95831dbd4ea   \n",
       "4  b6b74adb-31e0-4ecf-9a65-aa8a1babc46b   \n",
       "\n",
       "                                           File Name          Data Category  \\\n",
       "0  TCGA-LUAD.6ab7dce7-f54d-4f9e-b232-7f094e8bf05d...  Copy Number Variation   \n",
       "1  TCGA-LUAD.6e3bfcae-bea1-46b3-afee-df2e3a344cd1...  Copy Number Variation   \n",
       "2  TCGA-LUAD.1afb48ce-59a8-4130-8ff6-45d29405087e...  Copy Number Variation   \n",
       "3  TCGA-LUAD.d0597d14-df0c-413b-888f-53597d1fb61a...  Copy Number Variation   \n",
       "4  TCGA-LUAD.780c0b4e-e589-4cc3-b0da-ea99e908afdd...  Copy Number Variation   \n",
       "\n",
       "                Data Type Project ID       Case ID         Sample ID  \\\n",
       "0  Gene Level Copy Number  TCGA-LUAD  TCGA-38-4630  TCGA-38-4630-01A   \n",
       "1  Gene Level Copy Number  TCGA-LUAD  TCGA-50-5932  TCGA-50-5932-01A   \n",
       "2  Gene Level Copy Number  TCGA-LUAD  TCGA-J2-A4AD  TCGA-J2-A4AD-01A   \n",
       "3  Gene Level Copy Number  TCGA-LUAD  TCGA-38-4625  TCGA-38-4625-01A   \n",
       "4  Gene Level Copy Number  TCGA-LUAD  TCGA-55-7284  TCGA-55-7284-01B   \n",
       "\n",
       "     Sample Type  \n",
       "0  Primary Tumor  \n",
       "1  Primary Tumor  \n",
       "2  Primary Tumor  \n",
       "3  Primary Tumor  \n",
       "4  Primary Tumor  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove duplication cases\n",
    "df_cnv_sample_sheet = df_cnv_sample_sheet.drop_duplicates(subset=\"Case ID\", keep='first').reset_index(drop=True)\n",
    "df_cnv_sample_sheet.shape\n",
    "df_cnv_sample_sheet.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d6550153-50ad-43fe-b83c-60696a7849b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File ID</th>\n",
       "      <th>File Name</th>\n",
       "      <th>Case ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [File ID, File Name, Case ID]\n",
       "Index: []"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for duplication cases\n",
    "value_counts = df_cnv_sample_sheet['Case ID'].value_counts()\n",
    "df_cnv_sample_sheet[df_cnv_sample_sheet['Case ID'].isin(value_counts[value_counts > 1].index)][[\"File ID\", \"File Name\", \"Case ID\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1f20eff3-9132-4b06-b4dc-b98180bfb244",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(493, 8)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get cnv data of case exist in clinical data\n",
    "df_cnv_sample_sheet = df_cnv_sample_sheet[df_cnv_sample_sheet[\"Case ID\"].isin(list_cases)]\n",
    "df_cnv_sample_sheet.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02f90c23-84cc-447b-8f36-47249abdb2c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loop through all rows\n",
    "case_data = []\n",
    "if cnv_process_status != \"processed\":\n",
    "    for index, row in df_cnv_sample_sheet.iterrows():\n",
    "        cnv_data_file_path = f\"{cnv_data_dir}/{row['File ID']}/{row['File Name']}\"\n",
    "        case_id = row[\"Case ID\"]\n",
    "        # Check if file exist\n",
    "        if os.path.exists(cnv_data_file_path):\n",
    "            # logger.info(f\"Processing case: {row['Case ID']} at {cnv_data_file_path}\")\n",
    "            df_cnv_data = pd.read_csv(cnv_data_file_path, sep=\"\\t\")\n",
    "            # print(df_cnv_data)\n",
    "            df_cnv_data = df_cnv_data[df_cnv_data['copy_number'].notna()][['chromosome','gene_name', 'copy_number']]\n",
    "            df_cnv_data = df_cnv_data.groupby('gene_name', as_index=False)['copy_number'].mean()\n",
    "            \n",
    "            # value_counts = df_cnv_data['gene_name'].value_counts()\n",
    "            # print(df_cnv_data[df_cnv_data['gene_name'].isin(value_counts[value_counts > 1].index)][[\"gene_name\", \"copy_number\"]])\n",
    "            df_cnv_case_data = df_cnv_data.set_index('gene_name').T\n",
    "            df_cnv_case_data['case_id'] = case_id\n",
    "            df_cnv_case_data = df_cnv_case_data.set_index('case_id')\n",
    "       \n",
    "            case_data.append(df_cnv_case_data)\n",
    "        else:\n",
    "            logger.error(f\"File {cnv_data_file_path} not exists!\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d88c363-614e-4c14-8002-fba58de7aee6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge all cases\n",
    "if cnv_process_status != \"processed\":\n",
    "    df_cnv_data = pd.concat(case_data, axis=0)\n",
    "    df_cnv_data.columns.name = None\n",
    "    df_cnv_data.to_csv(f\"{processed_dir}/cnv_raw.tsv\", sep=\"\\t\")\n",
    "else:\n",
    "    df_cnv_data = pd.read_csv(f\"{processed_dir}/cnv_raw.tsv\", sep=\"\\t\")\n",
    "    \n",
    "logger.info(df_cnv_data.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f47ea6d5-c8cf-406f-a12b-d4f352978bce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the percentage of missing values per column\n",
    "missing_percentage = df_cnv_data.isnull().mean() * 100\n",
    "\n",
    "# Summarize the number of columns in different ranges of missing values\n",
    "summary = {\n",
    "    '0-10%': (missing_percentage <= 10).sum(),\n",
    "    '10-30%': ((missing_percentage > 10) & (missing_percentage <= 30)).sum(),\n",
    "    '30-50%': ((missing_percentage > 30) & (missing_percentage <= 50)).sum(),\n",
    "    '50-100%': (missing_percentage > 50).sum()\n",
    "}\n",
    "logger.info(\"Summary of missing value percentages:\")\n",
    "logger.info(summary)\n",
    "\n",
    "# Define a threshold for missing values\n",
    "threshold = 0.5  # 50% threshold\n",
    "\n",
    "# Drop columns with more than 50% missing values\n",
    "df_cnv_data_processed = df_cnv_data.loc[:, missing_percentage <= (threshold * 100)]\n",
    "logger.info(df_cnv_data_processed.shape)\n",
    "\n",
    "# Fill NA with CN = 2 which is normal for human\n",
    "df_cnv_data_processed = df_cnv_data_processed.fillna(2)\n",
    "\n",
    "# Add prefix for every field\n",
    "df_cnv_data_processed = df_cnv_data_processed.add_prefix(\"CNV_gene_\")\n",
    "\n",
    "logger.info(df_cnv_data_processed.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "127bbff0-8b7b-4864-9c82-476c825e29c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export data\n",
    "if cnv_process_status != \"processed\":\n",
    "    df_cnv_data_processed.to_csv(f\"{processed_dir}/cnv.tsv\", sep=\"\\t\")\n",
    "else:\n",
    "    df_cnv_data_processed = pd.read_csv(f\"{processed_dir}/cnv.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b145cfcc-ec9b-4f01-9cf8-31d3cbb34410",
   "metadata": {},
   "source": [
    "### Preprocess DNA Methylation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae85072e-8a5b-423d-a72e-17751d02c340",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load DNA methylation sample sheet\n",
    "df_dna_methylation_sample_sheet = pd.read_csv(f\"{sample_sheets_dir}/gdc_dna_methylation_sample_sheet.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "365cadbb-27b0-4200-b623-78b857f9d329",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check for duplication cases\n",
    "value_counts = df_dna_methylation_sample_sheet['Case ID'].value_counts()\n",
    "df_dna_methylation_sample_sheet[df_dna_methylation_sample_sheet['Case ID'].isin(value_counts[value_counts > 1].index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e723ca9-4377-4890-a85d-e1a564c03bc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove duplication cases\n",
    "logger.info(df_dna_methylation_sample_sheet.shape)\n",
    "\n",
    "# Keeping only Primary Tumor\n",
    "df_dna_methylation_sample_sheet = df_dna_methylation_sample_sheet[df_dna_methylation_sample_sheet['Sample Type'] == 'Primary Tumor']\n",
    "\n",
    "# Keep the first Case ID only\n",
    "df_dna_methylation_sample_sheet = df_dna_methylation_sample_sheet.drop_duplicates(subset=\"Case ID\", keep='first').reset_index(drop=True)\n",
    "\n",
    "logger.info(df_dna_methylation_sample_sheet.shape)\n",
    "df_dna_methylation_sample_sheet.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9dceb91-70fc-4c63-9d51-f66b4444ed5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loop through all rows\n",
    "dna_methylation_case_data = []\n",
    "dna_methylation_column_names = [\"id\", \"value\"]\n",
    "\n",
    "logger.info(df_dna_methylation_sample_sheet.shape)\n",
    "\n",
    "for index, row in df_dna_methylation_sample_sheet.iterrows():\n",
    "    dna_methylation_data_file_path = f\"{dna_methylation_data_dir}/{row['File ID']}/{row['File Name']}\"\n",
    "    case_id = row[\"Case ID\"]\n",
    "    logger.info(f\"At: {index}.\")\n",
    "    # Check if file exist\n",
    "    if os.path.exists(dna_methylation_data_file_path):\n",
    "        # logger.info(f\"Processing case: {row['Case ID']} at {dna_methylation_data_file_path}\")\n",
    "        df_dna_methylation_data = pd.read_csv(dna_methylation_data_file_path, header=None, names=dna_methylation_column_names, sep=\"\\t\")\n",
    "        \n",
    "        # Include only cg probes\n",
    "        df_dna_methylation_data = df_dna_methylation_data[df_dna_methylation_data['id'].astype(str).str.startswith('cg')]\n",
    "\n",
    "        # Exclude NaN value\n",
    "        df_dna_methylation_data = df_dna_methylation_data[df_dna_methylation_data['value'].notna()]\n",
    "    \n",
    "        df_dna_methylation_case_data = df_dna_methylation_data.set_index('id').T\n",
    "        df_dna_methylation_case_data['case_id'] = case_id\n",
    "        df_dna_methylation_case_data = df_dna_methylation_case_data.set_index('case_id')\n",
    "   \n",
    "        dna_methylation_case_data.append(df_dna_methylation_case_data)\n",
    "        \n",
    "        logger.info(f\"At: {index}. Shape: {df_dna_methylation_data.shape}\")\n",
    "    else:\n",
    "        logger.error(f\"File {cnv_data_file_path} not exists!\")\n",
    "logger.info(\"Done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf0875cb-9ca3-4aee-abf0-5949c71296b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge all cases\n",
    "if dna_methylation_process_status != \"processed\":\n",
    "    df_dna_methylation_data = pd.concat(dna_methylation_case_data, axis=0)\n",
    "    df_dna_methylation_data.columns.name = None\n",
    "    df_dna_methylation_data.to_csv(f\"{processed_dir}/dna_methylation_raw.tsv\", sep=\"\\t\")\n",
    "else:\n",
    "    df_dna_methylation_data = pd.read_csv(f\"{processed_dir}/dna_methylation_raw.tsv\", sep=\"\\t\")\n",
    "    \n",
    "logger.info(df_dna_methylation_data.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23ea627d-6099-46c9-baee-b1a1cbdfecb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the percentage of missing values per column\n",
    "missing_percentage = df_dna_methylation_data.isnull().mean() * 100\n",
    "\n",
    "# Summarize the number of columns in different ranges of missing values\n",
    "summary = {\n",
    "    '0-10%': (missing_percentage <= 10).sum(),\n",
    "    '10-30%': ((missing_percentage > 10) & (missing_percentage <= 30)).sum(),\n",
    "    '30-50%': ((missing_percentage > 30) & (missing_percentage <= 50)).sum(),\n",
    "    '50-100%': (missing_percentage > 50).sum()\n",
    "}\n",
    "logger.info(\"Summary of missing value percentages:\")\n",
    "logger.info(summary)\n",
    "\n",
    "# Define a threshold for missing values\n",
    "threshold = 0.2  # 20% threshold\n",
    "\n",
    "# Drop columns with more than 50% missing values\n",
    "df_dna_methylation_data_processed = df_dna_methylation_data.loc[:, missing_percentage <= (threshold * 100)]\n",
    "logger.info(df_dna_methylation_data_processed.shape)\n",
    "\n",
    "# Impute missing values with the mean of each probe\n",
    "df_dna_methylation_data_processed = df_dna_methylation_data_processed.fillna(df_dna_methylation_data_processed.mean())\n",
    "\n",
    "# Add prefix for every field\n",
    "df_dna_methylation_data_processed = df_dna_methylation_data_processed.add_prefix(\"DNA_meth_\")\n",
    "\n",
    "logger.info(df_dna_methylation_data_processed.head())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
